test/functional/logforwarding/json-parsing.go

```diff
@@ -0,0 +1,106 @@
+package logforwarding
+
+import (
+	"fmt"
+	"strings"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	logging "github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1"
+	"github.com/openshift/cluster-logging-operator/test/functional"
+	"github.com/openshift/cluster-logging-operator/test/helpers/types"
+	//	. "github.com/openshift/cluster-logging-operator/test/matchers"
+)
+
+var _ = Describe("[LogForwarding] Json log parsing", func() {
+	var (
+		framework *functional.FluentdFunctionalFramework
+		clfb      *functional.ClusterLogForwarderBuilder
+	)
+	BeforeEach(func() {
+		framework = functional.NewFluentdFunctionalFramework()
+		clfb = functional.NewClusterLogForwarderBuilder(framework.Forwarder).
+			FromInput(logging.InputNameApplication).
+			ToFluentForwardOutput()
+	})
+	AfterEach(func() {
+		framework.Cleanup()
+	})
+
+	It("should parse json message into structured", func() {
+		clfb.Forwarder.Spec.Pipelines[0].Parse = "json"
+		Expect(framework.Deploy()).To(BeNil())
+		// Log message data
+		jsonMsg := `
+		{
+			"name":"Benjamin Watson",
+			"age":31,
+			"isMarried":true,
+			"hobbies":["Football","Swimming"],
+			"kids":[
+			  {
+				  "name":"Billy",
+				  "age":5
+			  },
+			  {
+				  "name":"Milly",
+				  "age":3
+			  }
+			]
+		}`
+		expected := map[string]interface{}{
+			"name":      "Benjamin Watson",
+			"age":       float64(31),
+			"isMarried": true,
+			"hobbies":   []interface{}{"Football", "Swimming"},
+			"kids": []interface{}{
+				map[string]interface{}{
+					"name": "Billy",
+					"age":  float64(5),
+				},
+				map[string]interface{}{
+					"name": "Milly",
+					"age":  float64(3),
+				},
+			},
+		}
+		jsonMsg = strings.ReplaceAll(jsonMsg, "\n", "")
+		expectedMessage := jsonMsg
+		jsonMsg = strings.ReplaceAll(jsonMsg, "\"", "\\\"")
+		timestamp := "2020-11-04T18:13:59.061892+00:00"
+
+		// Write log line as input to fluentd
+		applicationLogLine := fmt.Sprintf("%s stdout F %s", timestamp, jsonMsg)
+		Expect(framework.WriteMessagesToApplicationLog(applicationLogLine, 10)).To(BeNil())
+		// Read line from Log Forward output
+		raw, err := framework.ReadApplicationLogsFrom("fluentforward")
+		Expect(err).To(BeNil(), "Expected no errors reading the logs")
+		// Parse log line
+		var logs []types.ApplicationLog
+		err = types.StrictlyParseLogs(raw, &logs)
+		Expect(logs[0].Structured).To(Equal(expected), "received structured not matching")
+		Expect(logs[0].Message).To(Equal(expectedMessage), "received message not matching")
```

Comment:
@alanconway please confirm.  i think the expectation is that "message" is empty:

```
For each log entry; if parse: json is set and the entry is valid JSON, the output record will have a structured field. Otherwise the output record will have a message field.
```

Reply:
As per conversation last week, this requirement is relaxed. Now `message` will remain as is. `structured` will be populated after parsing.